subsys/dfu/src/dfu_target.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <zephyr.h>
+#include <logging/log.h>
+#include <dfu/mcuboot.h>
+#include <dfu/dfu_target.h>
+#include "dfu_target_mcuboot.h"
+#include "dfu_target_modem.h"
+
+#ifdef CONFIG_DFU_TARGET_MODEM
+extern struct dfu_target dfu_target_modem;
```

Comment:
Since you have these structs, you don't need the individual header files.
Alternatively, you can move the struct declaration into this file and keep the h files. Either is fine with me.

Reply:
They are used for the `dfu_target_<type>_identify` functions.

Reply:
I do agree that we don't need to have individual header files for the rest of the functions, but isn't it nice to keep as documentation or do you think the maintenance cost is too high? They are used internally by `dfu_target_<type>.c` files for forward declarations.

Reply:
Other motivations for having the header files are:
- Documentation: some of these target have strange requirements, which explain why the dfu_target is used how it is
- "static" (as in not belonging to a class in a OO language) functions will soon be added to these DFU targets, and they need to be exposed through these header files. 

I do however see that one might want to simplify things and leave the functions which are exposed in the struct out of the header file. The docs for the shared dfu target functions could be in the .c file also.

Reply:
Ok, it's been argued before, I won't request further changes this late. Since I like a good discussion, though, here are my thoughts, take it or leave it ;):

I agree the header files are good, they are a nice, familiar way to present the API. The way it is now, though, the header files are buried very deep from the user's perspective. To find the documentation for the API they are given, they must

1. find which struct they have been given, then
2. (find the extern declaration, then)
3. find the real declaration and initialization, then
4. find the header file.

My suggestion would be to have the struct declaration and initialization in the dfu_target.c file, which has the added benefit of no circular dependency (dfu_target_xxx.c currently `#includes` dfu_target.h).

Reply:
This iteration has also been done at some point :+1: but I'm not opposed to the idea.